src/liballoc/rc.rs

```diff
@@ -1708,9 +1702,18 @@ impl<T> Weak<T> {
     /// [`null`]: ../../std/ptr/fn.null.html
     #[stable(feature = "weak_into_raw", since = "1.45.0")]
     pub fn as_ptr(&self) -> *const T {
-        let offset = data_offset_sized::<T>();
-        let ptr = self.ptr.cast::<u8>().as_ptr().wrapping_offset(offset);
-        ptr as *const T
+        let ptr: *mut RcBox<T> = NonNull::as_ptr(self.ptr);
+
+        // SAFETY: we must offset the pointer manually, and said pointer may be
+        // a dangling weak (usize::MAX). data_offset is safe to call, because we
```

Comment:
My concern here has not been resolved. My impression is that this would be easier to understand if the very first sentence said something like

>  SAFETY: we must offset the pointer manually, and said pointer may be a dangling weak (usize::MAX) -- but only if T is sized.

In some way, the fact that dangling and sizedness are connected should be mentioned the moment either of these two concepts is introduced; right now it is a rather surprising connection that is only established after the fact.

I might be nitpicking, but I feel clarity is really important here. So basically if anything is even slightly confusing  on the first read, I try to improve it.^^

Reply:
There, I think that addresses it.